File: backend/analytics.py
"""Shared training-analytics helpers.

Pure functions with no DB or framework dependencies so they're trivial to unit-test
and reusable across the suggestion engine, progress charts, PR detection, and the
plateau detector.
"""
from __future__ import annotations

from datetime import date
from typing import Optional, Sequence, Tuple


def e1rm(weight: Optional[float], reps: Optional[int]) -> Optional[float]:
    """Estimated 1-rep max via the Epley formula: weight * (1 + reps / 30).

    Returns None if inputs are missing. A 1-rep set returns the weight itself.
    e1RM normalizes strength across rep ranges so sessions with different
    weight x rep schemes can be compared on a single axis.
    """
    if weight is None or reps is None or weight <= 0 or reps <= 0:
        return None
    return round(weight * (1 + reps / 30), 2)


def best_e1rm(sets) -> Optional[float]:
    """Highest e1RM across a collection of SetLog-like objects.

    Each set exposes `weight_kg` and `reps` (one load per row, regardless of side).
    Warm-ups are skipped when the flag is present; lighter sets wouldn't win anyway,
    but excluding them keeps the semantics honest.
    """
    best = None
    for s in sets:
        if getattr(s, "is_warmup", False):
            continue
        est = e1rm(getattr(s, "weight_kg", None), getattr(s, "reps", None))
        if est is not None and (best is None or est > best):
            best = est
    return best


def double_progression_cue(sets, target_reps_max: Optional[int]) -> Optional[dict]:
    """Double-progression signal from last session's working sets.

    Needs a rep-range ceiling (target_reps_max) to judge against. Returns:
      - level "add": you hit the top of the range on at least one set → add weight.
      - level "try": you didn't hit it, but reps + RIR says you had it in you
        (e.g. 7 reps @ 1 RIR or 6 reps @ 2 RIR against a max of 8) → try adding weight.
      - None: no rep ceiling, no working sets, or still mid-range.
    """
    if not target_reps_max:
        return None
    working = [s for s in sets if not getattr(s, "is_warmup", False) and s.rir is not None]
    if not working:
        return None

    if any(s.reps >= target_reps_max for s in working):
        return {
            "level": "add",
            "message": f"Add weight — you hit the top of your rep range ({target_reps_max}) last time.",
        }
    if any((s.reps + s.rir) >= target_reps_max for s in working):
        return {
            "level": "try",
            "message": f"Try adding weight — you had reps in reserve at the top of your range ({target_reps_max}) last time.",
        }
    return None


def linreg_slope(xs: Sequence[float], ys: Sequence[float]) -> Optional[float]:
    """Slope of the least-squares best-fit line through (xs, ys).

    Returns the change in y per unit x, or None if it can't be computed (fewer
    than 2 points, or all x identical). This is the plain closed-form formula:
    slope = Σ(x-x̄)(y-ȳ) / Σ(x-x̄)².
    """
    n = len(xs)
    if n < 2:
        return None
    mean_x = sum(xs) / n
    mean_y = sum(ys) / n
    num = sum((x - mean_x) * (y - mean_y) for x, y in zip(xs, ys))
    den = sum((x - mean_x) ** 2 for x in xs)
    if den == 0:
        return None
    return num / den


def detect_plateau(
    series: Sequence[Tuple[date, Optional[float]]],
    min_sessions: int = 4,
    window_weeks: int = 4,
    min_gain_per_week: float = 0.5,
) -> dict:
    """Decide whether an exercise's e1RM trend has stalled.

    `series` is (date, e1rm) sorted ascending. We fit a line through the recent
    window and read its slope in kg/week: clearly rising → progressing, flat or
    falling → plateaued. Thresholds are parameters, not truths — tune to taste.
    """
    pts = [(d, v) for d, v in series if v is not None]
    insufficient = {
        "is_plateaued": False,
        "status": "insufficient_data",
        "message": "Not enough history yet to judge a trend.",
        "slope_per_week": None,
        "weeks_analyzed": 0.0,
    }
    if len(pts) < min_sessions:
        return insufficient

    # Prefer points inside the recent window; fall back to the last min_sessions.
    last_date = pts[-1][0]
    window = [(d, v) for d, v in pts if (last_date - d).days <= window_weeks * 7]
    if len(window) < min_sessions:
        window = pts[-min_sessions:]

    span_weeks = (window[-1][0] - window[0][0]).days / 7
    if span_weeks < 2:
        return insufficient

    days = [(d - window[0][0]).days for d, _ in window]
    vals = [v for _, v in window]
    slope_day = linreg_slope(days, vals)
    if slope_day is None:
        return insufficient

    slope_per_week = round(slope_day * 7, 2)
    is_plateaued = slope_per_week < min_gain_per_week
    if is_plateaued:
        message = (
            f"e1RM has moved {slope_per_week:+.1f} kg/week over the last "
            f"{round(span_weeks, 1)} weeks — progress has stalled."
        )
    else:
        message = f"Progressing at {slope_per_week:+.1f} kg/week."

    return {
        "is_plateaued": is_plateaued,
        "status": "ok",
        "message": message,
        "slope_per_week": slope_per_week,
        "weeks_analyzed": round(span_weeks, 1),
    }

File: frontend/src/local/analytics.js
// Pure training-analytics helpers — a faithful JS port of the backend analytics.py
// plus the suggestion engine from sessions.py. No DB access here.

const round2 = (n) => Math.round(n * 100) / 100

// Local calendar date (YYYY-MM-DD), matching how the app stores dates.
export function todayLocal() {
  return new Date().toLocaleDateString('en-CA')
}

// Parse a YYYY-MM-DD string at noon to dodge timezone date-rollover.
export function parseDate(d) {
  return new Date(d + 'T12:00:00')
}

function daysBetween(a, b) {
  return Math.round((parseDate(b) - parseDate(a)) / 86400000)
}

// Epley: weight * (1 + reps/30). null on missing/non-positive inputs.
export function e1rm(weight, reps) {
  if (weight == null || reps == null || weight <= 0 || reps <= 0) return null
  return round2(weight * (1 + reps / 30))
}

export function bestE1rm(sets) {
  let best = null
  for (const s of sets) {
    if (s.is_warmup) continue
    const est = e1rm(s.weight_kg, s.reps)
    if (est != null && (best == null || est > best)) best = est
  }
  return best
}

export function doubleProgressionCue(sets, targetRepsMax) {
  if (!targetRepsMax) return null
  const working = sets.filter(s => !s.is_warmup && s.rir != null)
  if (!working.length) return null

  if (working.some(s => s.reps >= targetRepsMax)) {
    return { level: 'add', message: `Add weight next session!! You hit the top of your rep range(${targetRepsMax}) last time.` }
  }
  if (working.some(s => (s.reps + s.rir) >= targetRepsMax)) {
    return { level: 'try', message: `Try adding weight — you had reps in reserve at the top of your range (${targetRepsMax}) last time.` }
  }
  return null
}

function linregSlope(xs, ys) {
  const n = xs.length
  if (n < 2) return null
  const meanX = xs.reduce((a, b) => a + b, 0) / n
  const meanY = ys.reduce((a, b) => a + b, 0) / n
  let num = 0, den = 0
  for (let i = 0; i < n; i++) {
    num += (xs[i] - meanX) * (ys[i] - meanY)
    den += (xs[i] - meanX) ** 2
  }
  if (den === 0) return null
  return num / den
}

// series: array of [dateStr, e1rm|null] sorted ascending by date.
export function detectPlateau(series, { minSessions = 4, windowWeeks = 4, minGainPerWeek = 0.5 } = {}) {
  const pts = series.filter(([, v]) => v != null)
  const insufficient = {
    is_plateaued: false,
    status: 'insufficient_data',
    message: 'Not enough history yet to judge a trend.',
    slope_per_week: null,
    weeks_analyzed: 0.0,
  }
  if (pts.length < minSessions) return insufficient

  const lastDate = pts[pts.length - 1][0]
  let window = pts.filter(([d]) => daysBetween(d, lastDate) <= windowWeeks * 7)
  if (window.length < minSessions) window = pts.slice(-minSessions)

  const spanWeeks = daysBetween(window[0][0], window[window.length - 1][0]) / 7
  if (spanWeeks < 2) return insufficient

  const days = window.map(([d]) => daysBetween(window[0][0], d))
  const vals = window.map(([, v]) => v)
  const slopeDay = linregSlope(days, vals)
  if (slopeDay == null) return insufficient

  const slopePerWeek = round2(slopeDay * 7)
  const isPlateaued = slopePerWeek < minGainPerWeek
  const sign = (n) => (n >= 0 ? `+${n.toFixed(1)}` : n.toFixed(1))
  const message = isPlateaued
    ? `e1RM has moved ${sign(slopePerWeek)} kg/week over the last ${round2(spanWeeks).toFixed(1)} weeks — progress has stalled.`
    : `Progressing at ${sign(slopePerWeek)} kg/week.`

  return {
    is_plateaued: isPlateaued,
    status: 'ok',
    message,
    slope_per_week: slopePerWeek,
    weeks_analyzed: round2(spanWeeks),
  }
}

// Weight suggestion from last session's sets — ported from compute_suggestion().
export function computeSuggestion(exercise, lastSets) {
  const increments = { push: 2.5, pull: 2.5, squat: 5.0, hinge: 5.0, carry: 2.5 }
  let step = increments[exercise.movement_pattern] ?? 2.5
  if (exercise.is_unilateral && (exercise.movement_pattern === 'push' || exercise.movement_pattern === 'pull')) {
    step = 1.25
  }

  const working = lastSets.filter(s => s.rir != null && !s.is_warmup)
  if (!working.length) {
    return { reason: 'No sets logged yet', weight: null, weight_left: null, weight_right: null }
  }

  const avgRir = working.reduce((a, s) => a + s.rir, 0) / working.length
  let action, weightChange
  if (avgRir >= 3) { action = 'Add weight — too easy'; weightChange = step }
  else if (avgRir >= 1) { action = 'Same weight, aim for more reps'; weightChange = 0 }
  else { action = 'Consolidate before progressing'; weightChange = 0 }

  const lastSideWeight = (side) => {
    const sideSets = working.filter(s => s.side === side)
    if (!sideSets.length) return null
    return round2((sideSets[sideSets.length - 1].weight_kg || 0) + weightChange)
  }

  if (exercise.is_unilateral) {
    return { reason: action, weight: null, weight_left: lastSideWeight('left'), weight_right: lastSideWeight('right') }
  }
  const lastWeight = working[working.length - 1].weight_kg || 0
  return { reason: action, weight: round2(lastWeight + weightChange), weight_left: null, weight_right: null }
}

// Python's date.weekday(): Mon=0..Sun=6. JS getDay(): Sun=0..Sat=6.
export function pyWeekday(dateStr) {
  return (parseDate(dateStr).getDay() + 6) % 7
}
